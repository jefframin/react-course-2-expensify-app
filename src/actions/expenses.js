import uuid from 'uuid'
import database from '../firebase/firebase'

export const addExpenseAction = ({ id, description, note, amount, created } = {}) => {
    return {
        type: 'ADD_EXPENSE',
        expense: {
        id: id,
        description: description,
        note: note,
        amount: amount,
        created: created
        }
    }
}

export const removeExpenseAction = ({ id } = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    }
}

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpenseAction({ id }))
        })
    };
}

export const editExpenseAction = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates
    }
}

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note =  '',
            amount = 0,
            created = 0
        } = expenseData
        const expense = { description, note, amount, created }
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpenseAction({
                id: ref.key,
                ...expense
            }))
        })
    };
}

// set expenses
export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses: expenses
    }
} 
        
export const startSetExpenses = () => {

    return (dispatch) => {

        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = []
            snapshot.forEach((element) => {
                expenses.push({
                    ...element.val(),
                    id: element.key
                });
            });
        dispatch(setExpenses(expenses));
    });
};
};