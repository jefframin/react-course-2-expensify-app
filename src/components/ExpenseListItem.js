import React from 'react'
import { Link } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, created}) => {
    return (
        <div>
          <h3>
          <Link to={`/edit/${id}`}>{description}</Link>
          </h3>
          <p>{amount} - {created}</p>
        </div>
    )
}

export default ExpenseListItem