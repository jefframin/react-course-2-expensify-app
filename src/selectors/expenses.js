import moment from 'moment'

const getVisibleExpenses = ( {expenses, filters} ) => {
    return expenses.filter((expense) => {
            const timeCreated = moment(expense.created)
            return (expense.description.toLowerCase().includes(filters.text.toLowerCase()) &&
                   (filters.startDate ? filters.startDate.isSameOrBefore(timeCreated, 'day') : true) &&
                   (filters.endDate ? filters.endDate.isSameOrAfter(timeCreated, 'day') : true))
        }).sort((a, b) => {
            switch (filters.order) {
                case 'amount':
                    return a.amount < b.amount ? -1 : 1
                case 'date':
                    return a.created < b.created ? -1 : 1
                default:
                    return 0
            }
        })
}

export default getVisibleExpenses