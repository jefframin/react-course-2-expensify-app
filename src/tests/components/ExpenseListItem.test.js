import React from 'react'
import { shallow } from 'enzyme'
import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

//ExpenseListItem = ({id, description, amount, created}) 
test('should render expense list item with expense', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot()
})