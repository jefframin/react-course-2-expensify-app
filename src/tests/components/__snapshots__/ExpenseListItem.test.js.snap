// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item with expense 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="/edit/1"
    >
      gum
    </Link>
  </h3>
  <p>
    $1.95
    -
    December 31st, 1969
  </p>
</div>
`;
